pkitesting/src/main/java/io/netty/pkitesting/RevocationServer.java

```diff
@@ -0,0 +1,188 @@
+                throw new IllegalStateException(e);
+            }
+        }
+    }
+
+    /**
+     * Register an issuer with the revocation server.
+     * This must be done before CRLs can be served for that issuer, and before any of its certificates can be revoked.
+     * @param issuer The issuer to register.
+     */
+    public void register(X509Bundle issuer) {
+        issuers.computeIfAbsent(issuer.getCertificate(), bundle -> {
+            String path = "/crl/" + issuerCounter.incrementAndGet() + ".crl";
+            URI uri = URI.create(crlBaseAddress + path);
+            CrlInfo info = new CrlInfo(issuer, uri);
+            paths.put(path, info);
+            return info;
+        });
+    }
+
+    /**
+     * Revoke the given certificate with the given revocation time.
+     * <p>
+     * The issuer of the given certificate must be {@linkplain #register(X509Bundle) registered} before its certifiactes
+     * can be revoked.
+     * @param cert The certificate to revoke.
+     * @param time The time of revocation.
+     */
+    public void revoke(X509Bundle cert, Instant time) {
+        X509Certificate[] certPath = cert.getCertificatePathWithRoot();
+        X509Certificate issuer = certPath.length == 1 ? certPath[0] : certPath[1];
+        CrlInfo info = issuers.get(issuer);
+        if (info != null) {
+            info.revokedCerts.put(cert.getCertificate().getSerialNumber(), time);
+        } else {
+            throw new IllegalArgumentException("Not a registered issuer: " + issuer.getSubjectX500Principal());
+        }
+    }
+
+    /**
+     * Get the URI of the Certificate Revocation List for the given issuer.
+     * @param issuer The issuer to get the CRL for.
+     * @return The URI to the CRL for the given issuer,
+     * or {@code null} if the issuer is not {@linkplain #register(X509Bundle) registered}.
+     */
+    public URI getCrlUri(X509Bundle issuer) {
+        CrlInfo info = issuers.get(issuer.getCertificate());
+        if (info != null) {
+            return info.uri;
+        }
+        return null;
+    }
+
+    private static byte[] generateCrl(CrlInfo info) {
+        X509Bundle issuer = info.issuer;
+        Map<BigInteger, Instant> certs = info.revokedCerts;
+        Instant now = Instant.now();
+        CertificateList list = new CertificateList(issuer, now, now, certs.entrySet());
+        try {
+            Signed signed = new Signed(list::getEncoded, issuer);
+            return signed.getEncoded();
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to sign CRL", e);
+        }
+    }
+
+    private static final class CrlInfo {
+        private final X509Bundle issuer;
+        private final URI uri;
+        private final Map<BigInteger, Instant> revokedCerts;
+
+        CrlInfo(X509Bundle issuer, URI uri) {
+            this.issuer = issuer;
+            this.uri = uri;
+            revokedCerts = new ConcurrentHashMap<>();
```

Comment:
Do we need to make this bounded ?

Reply:
Certificate revokation is forever.

I think this is fine because the revocation server is only for intended for testing usage. A real CA would maintain this information in a database.